src/python/Utils/ExecuteDAO.py

```diff
@@ -0,0 +1,217 @@
+    """
+    Return a logger which writes everything to stdout.
+    """
+    logger = logging.getLogger()
+    outHandler = logging.StreamHandler(sys.stdout)
+    outHandler.setFormatter(logging.Formatter("%(asctime)s:%(levelname)s:%(module)s: %(message)s"))
+    outHandler.setLevel(logLevel)
+    if logger.handlers:
+        logger.handlers.clear()
+    logger.addHandler(outHandler)
+    logger.setLevel(logLevel)
+    return logger
+
+
+class ExecuteDAO():
+    """
+    A generic class to create the DAO Factory and execute the DAO module.
+    """
+    def __init__(self, logger, package=None, daoModule=None):
+        """
+        __init__
+        The ExecuteDAO constructor method.
+        :param logger: The logger instance.
+        :param package: The Package from which the DAO factory to be initialised.
+        :param module: The DAO module to be executed.
+        """
+        # Connecting to DB
+        myThread = threading.currentThread()
+        connectToDB()
+        self.logger = logger
+        self.dbi = myThread.dbi
+        self.package = package
+        self.daoModule = daoModule
+        self.sqlRegEx = re.compile(".*sql.*", re.IGNORECASE)
+
+        self.daoFactory = DAOFactory(package=package,
+                                     logger=self.logger,
+                                     dbinterface=self.dbi)
+        self.logger.info("DAO Factory initialised from package: %s", self.package)
+
+        self.dao = self.daoFactory(classname=daoModule)
+        self.logger.info("DAO Module initialised as: %s", self.daoModule)
+
+
+    def __call__(self, dryRun=False, *sqlArgs, **sqlKwArgs):
+        """
+        __call__
+        The ExecuteDAO call method. This is the method to forward all provided
+        arguments to the execute method of the DAO and return the result from the query
+        :param dryRun:      Bool flag to indicate dryrun method
+        :param *sqlArgs:    All positional arguments to be forwarded to the DAO's execute method.
+        :param **sqlKwArgs: All named arguments to be forwarded to the DAO's execute method.
+        :return:            The result from the DAO. Depending on the DAO itself it Could be one of:
+                            * A dictionary
+                            * A list
+                            * A generator
+        """
+        if dryRun:
+            results = []
+            self.getHelp()
+            self.logger.info("DAO SQL queries to be executed:")
+            for sqlStr in self.getSqlQuery():
+                msg = "\n----------------------------------------------------------------------\n"
+                msg += "%s"
+                msg += "\n----------------------------------------------------------------------\n"
+                self.logger.info(msg, sqlStr)
+            self.logger.info("DAO SQL arguments provided:\n{}, {}".format(pformat(sqlArgs), pformat(sqlKwArgs)))
+        else:
+            results = self.dao.execute(*sqlArgs, **sqlKwArgs)
```

Comment:
@todor-ivanov I don't understand why you still claiming that return results can be a dict, a list a generator. You have full control in this function, so please convert to ONE and only one data-type, e.g. to the list of dict objects. The DAO usually return a dict since it represent a table. Now, you may check if results here is a list, a list of generator. Then make appropriate action to convert to only one single data type.

Reply:
Hi @vkuznet 

> I don't understand why you still claiming that return results can be a dict, a list a generator.

What I mean is it is all the DAOs that return different type of formats. Not this method alone. Here are some examples:

Here is one which is returning a dictionary having the workflow names as keys:
```
cmst1@vocms0193:/data/srv/wmagent $ python3 ExecuteDAO.py -c current/config/wmagent/config.py -p WMCore.WMBS -m Workflow.GetDeletableWorkflows
2022-06-13 13:50:50,159:INFO:DBCore: Instantiating base WM DBInterface
2022-06-13 13:50:50,160:INFO:ExecuteDAO: DAO Factory initialised from package: WMCore.WMBS
2022-06-13 13:50:50,162:INFO:ExecuteDAO: DAO Module initialised as: Workflow.GetDeletableWorkflows
2022-06-13 13:50:50,675:INFO:ExecuteDAO: DAO Results:
{'amaltaro_ACDC_02791_TC_6Tasks_PU_HG2206_Val_220607_161640_2668': {'spec': '/data/srv/wmagent/v2.0.2.patch3/install/wmagentpy3/WorkQueueManager/cache/amaltaro_ACDC_02791_TC_6Tasks_PU_HG2206_Val_220607_161640_2668/WMSandbox/WMWorkload.pkl',
                                                                    'workflows': {870: [1149],
                                                                                  871: [1150],
                                                                                  872: [1151],
                                                                                  873: [1152],
                                                                                  874: [1153],
                                                                                  875: [1154],
                                                                                  876: [1155],
                                                                                  877: [1156],
                                                                                  878: [1157],
                                                                                  879: [1158],
                                                                                  880: [1159],
                                                                                  881: [1160],
                                                                                  882: [1161],
                                                                                  883: [1162],
                                                                                  884: [1163],
                                                                                  885: [1164]}},
 'amaltaro_ACDC_02812_TC_6Tasks_PU_HG2206_Val_220607_161553_411': {'spec': '/data/srv/wmagent/v2.0.2.patch3/install/wmagentpy3/WorkQueueManager/cache/amaltaro_ACDC_02812_TC_6Tasks_PU_HG2206_Val_220607_161553_411/WMSandbox/WMWorkload.pkl',
                                                                   'workflows': {886: [1165],
                                                                                 887: [1166],
                                                                                 888: [1167],
                                                                                 889: [1168],
                                                                                 890: [1169],
                                                                                 891: [1170],
                                                                                 892: [1171],
                                                                                 893: [1172],
                                                                                 894: [1173],
                                                                                 895: [1174],
                                                                                 896: [1175],
                                                                                 897: [1176],
                                                                                 898: [1177],
                                                                                 899: [1178],
                                                                                 900: [1179],
                                                                                 901: [1180],
                                                                                 902: [1181],
                                                                                 903: [1182],
                                                                                 904: [1183]}},
...
}
```

Here is another one returning a generator out of a list of dictionaries - one per record in the database, having the column names as keys for the dictionary(it is converted to a list here just for the pretty printing):

```
cmst1@vocms0193:/data/srv/wmagent $ python3 ExecuteDAO.py -c current/config/wmagent/config.py -p WMCore.WMBS -m Workflow.GetDeletedBlocksByWorkflow
2022-06-13 13:50:15,078:INFO:DBCore: Instantiating base WM DBInterface
2022-06-13 13:50:15,079:INFO:ExecuteDAO: DAO Factory initialised from package: WMCore.WMBS
2022-06-13 13:50:15,081:INFO:ExecuteDAO: DAO Module initialised as: Workflow.GetDeletedBlocksByWorkflow
2022-06-13 13:50:15,834:INFO:ExecuteDAO: DAO Results:
[{'blocksDeleted': [],
  'blocksNotDeleted': ['/VBF_HHTo2G2Tau_CV_1_C2V_1_C3_0_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16MiniAODAPVv2-HIG_RunIISummer20UL16MiniAODAPVv2_02457_0_TC_6Tasks_PU_HG2206_Val_Alanv1-v11/MINIAODSIM#48d248e6-84ff-4bb8-b25e-82ef8f92106d',
                       '/VBF_HHTo2G2Tau_CV_1_C2V_1_C3_0_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16RECOAPV-HIG_RunIISummer20UL16RECOAPV_02812_0_TC_6Tasks_PU_HG2206_Val_Alanv1-v11/AODSIM#5aa47cee-45e1-4920-8e91-6b9ad05af1c2'],
  'name': 'amaltaro_ACDC_02791_TC_6Tasks_PU_HG2206_Val_220607_161640_2668'},
 {'blocksDeleted': [],
  'blocksNotDeleted': ['/b_bbar_4l_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16HLTAPV-myTask4_TaskChain_Prod_May2022_Val_Alanv15-v14/GEN-SIM-RAW#a6587b1e-5a8c-4492-8563-89c84df75933',
                       '/b_bbar_4l_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16wmLHEGENAPV-myTask1_TaskChain_Prod_May2022_Val_Alanv15-v11/LHE#1d79c8ab-0692-458d-9a89-879d58b16a68',
                       '/b_bbar_4l_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16wmLHEGENAPV-myTask1_TaskChain_Prod_May2022_Val_Alanv15-v11/GEN#8e2c867b-7117-46ee-911c-582dd28c4038',
                       '/b_bbar_4l_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16MiniAODAPV-myTask6_TaskChain_Prod_May2022_Val_Alanv15-v16/MINIAODSIM#9bfddce8-29f8-43b9-99d9-690e3810162a'],
  'name': 'amaltaro_TaskChain_Prod_May2022_Val_220520_123843_5700'},
 {'blocksDeleted': [],
  'blocksNotDeleted': ['/b_bbar_4l_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16MiniAODAPV-myTask6_TaskChain_Prod_HG2206_Val_Alanv11-v16/MINIAODSIM#7390691e-4f58-4ea9-b807-fb0b37104bd6',
                       '/b_bbar_4l_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16wmLHEGENAPV-myTask1_TaskChain_Prod_HG2206_Val_Alanv11-v11/LHE#ea3a24bc-c218-4a44-b87d-be46516dba62',
                       '/b_bbar_4l_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16HLTAPV-myTask4_TaskChain_Prod_HG2206_Val_Alanv11-v14/GEN-SIM-RAW#a6ddc3f5-b0db-4546-8ed4-3d5d0fdca196',
                       '/b_bbar_4l_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL16wmLHEGENAPV-myTask1_TaskChain_Prod_HG2206_Val_Alanv11-v11/GEN#75ece353-932e-428c-9a55-9b3b955c4f0b'],
  'name': 'amaltaro_TaskChain_Prod_HG2206_Val_220606_201532_757'},
...
]
```

Here is another one returning just a list:

```
cmst1@vocms0193:/data/srv/wmagent $ python3 ExecuteDAO.py -c current/config/wmagent/config.py -p WMComponent.DBS3Buffer -m GetCompletedWorkflows
2022-06-13 13:54:04,031:INFO:DBCore: Instantiating base WM DBInterface
2022-06-13 13:54:04,031:INFO:ExecuteDAO: DAO Factory initialised from package: WMComponent.DBS3Buffer
2022-06-13 13:54:04,057:INFO:ExecuteDAO: DAO Module initialised as: GetCompletedWorkflows
2022-06-13 13:54:04,513:INFO:ExecuteDAO: DAO Results:
['amaltaro_SC_LHE_Ext_May2022_Val_220520_123825_8709',
 'amaltaro_TaskChain_MC_HG2206_Val_220606_201530_3078',
 'amaltaro_SC_LHE_Ext_HG2206_Val_220606_201513_5737',
 'amaltaro_ACDC_mergeRECO_TC_EL8_HG2206_Val_220607_160446_4513',
...
]
```

I know it is on us to decide to which format we should stick for all the DAOs to be returned. But I do not think this method here is the place to convert the output. Here we are talking about a method which is supposed to be visualizing the output of the DAOs we are working with in order to inspect them, rather than trying to recognize which would be the data format returned by the DAO and make the output uniform here, risking to loose some information also. This would be quite of a long shot, unless all the DAOs we are using are not reworked to follow the same pattern/template (to which idea, I need to stress, I am a huge supporter, but this is another and bigger effort). 

If on the other hand we decide to use this as a wrapper class and achieve this very same uniformity in one go ... by making all the calls to the DAO factory through this utility ... Well I can agree, this would work as  well. It is a matter of agreement on us how to proceed here. I am good both ways.But if it is to be loft just as a debugging tool.. then we better not tough the ouput format but, rahter provide it to the user as it would be provided to the code of the calling module by the DAO itself.

Reply:
Thanks @vkuznet, as it was discussed during the WMCore meeting I suggest we leave the end type untouched in this script and use it to return the DAO output  as it is for debugging purposes, while we create a separate issue to achieve the following two goals:
* Create some kind of a standard/template in order to enforce to some extent uniformity in the way we define all our DAOs
* Set a wrapper class to properly convert the output of all of them to a list of dictionaries with key names set to match the column names